src/python/pants/engine/legacy/graph.py

```diff
@@ -543,6 +552,23 @@ def hydrate_sources(
   yield HydratedField(sources_field.arg, fileset_with_spec)
 
 
+@rule
+def hydrate_urls(urls_field: UrlsField) -> HydratedField:
+  """Given a UrlsField, fetch all urls and merge into a single Snapshot."""
+  spec_path = urls_field.address.spec_path
+
+  snapshots = yield [Get(Snapshot, UrlToFetch, u) for u in urls_field.urls]
+  relative_snapshot = yield Get(Snapshot, DirectoriesToMerge(tuple(s.directory_digest for s in snapshots)))
+  snapshot = yield Get(Snapshot, DirectoryWithPrefixToAdd(relative_snapshot.directory_digest, spec_path))
+
+  fileset = _eager_fileset_with_spec('', {'globs': snapshot.files}, snapshot)
```

Comment:
I'm not convinced pretending that these were globs is super useful here. AFAIR this is only used for debug/error printing, so having the original URLs would probably be useful for that case. Maybe FilesetWithSpec is an abstraction we actually need to change/kill here?

Reply:
Most definitely: it exists for v1 though. But yea, I suppose that we could lie about the globs.